lib/elixir/lib/map.ex

```diff
@@ -93,11 +96,10 @@ defmodule Map do
   """
   @spec new(Enum.t, (term -> {key, value})) :: map
   def new(enumerable, transform) do
-    fun = fn el, acc ->
-      {k, v} = transform.(el)
-      put(acc, k, v)
-    end
-    Enum.reduce(enumerable, %{}, fun)
+    enumerable
+    |> Enum.reduce([], &[transform.(&1) | &2])
+    |> :lists.reverse
+    |> :maps.from_list
```

Comment:
So, I think we have another reason to avoid Enum.into in map: It's nicer to optimize Enum.into in _terms_ of Map.new, rather than the other way around. For example. here's my proposed change to Enum.into for empty maps:

```
def into(enumerable, %{} = map) when map_size(map) == 0 do
  Map.new(enumerable)
end
```

Alternatively, we could have Enum.into do what my Map.new does now, and have Map.new use Enum.into. Thoughts?